confluent_kafka/schema_registry/protobuf.py

```diff
@@ -0,0 +1,437 @@
+        self.msg_index = _create_msg_index(descriptor)
+        self._schema = Schema(_schema_to_str(proto_file),
+                              schema_type='PROTOBUF')
+
+    @staticmethod
+    def _encode_index(buf, index):
+        """
+        Writes the index to buf formatted as unsigned varints.
+
+        Args:
+            buf (BytesIO): buffer to write index to
+            index [index]: Message index
+
+        Returns:
+            [bytes]: Serialized Message index
+
+        """
+        for value in index:
+            while (value & 0x7f) != 0:
+                buf.write(_bytes((value & 0x7f) | 0x80))
+                value >>= 7
+            buf.write(_bytes(value))
+
+    def resolve_dependencies(self, ctx, file_desc):
+        """
+        Resolves and optionally registers schema references recursively.
+
+        Args:
+            ctx (SerializationContext): Serialization context.
+
+            file_desc (FileDescriptor): file descriptor to traverse.
+
+        """
+        schema_refs = []
+        for dep in file_desc.dependencies:
+            dep_refs = self.resolve_dependencies(ctx, dep)
+            subject = self._ref_reference_subject_func(ctx, dep)
+            schema = Schema(_schema_to_str(dep),
+                            references=dep_refs,
+                            schema_type='PROTOBUF')
+            if self._auto_register:
+                self._registry.register_schema(subject, schema)
+
+            reference = self._registry.lookup_schema(subject, schema)
+            # schema_refs are per file descriptor
+            schema_refs.append(SchemaReference(dep.name,
+                                               subject,
+                                               reference.version))
+        return schema_refs
+
+    def __call__(self, msg, ctx):
+        """
+        Serializes a Protobuf Message to the Confluent Schema Registry
+        Protobuf binary format.
+
+        Args:
+            msg (Message): Protobuf message instance.
+
+            ctx (SerializationContext): Metadata pertaining to the serialization
+                operation.
+
+        Note:
+            None objects are represented as Kafka Null.
+
+        Raises:
+            SerializerError if any error occurs serializing obj
+
+        Returns:
+            bytes: Confluent Schema Registry formatted Protobuf bytes
+
+        """
+        if msg is None:
+            return None
+
+        subject = self._subject_name_func(ctx, msg.DESCRIPTOR.full_name)
+
+        if subject not in self._known_subjects:
+            self._schema.references = self.resolve_dependencies(ctx, msg.DESCRIPTOR.file)
+
+        self._schema_id = self._registry.register_schema(subject, self._schema)
+
+        with _ContextStringIO() as fo:
+            # Write the magic byte and schema ID in network byte order (big endian)
+            fo.write(struct.pack('>bI', _MAGIC_BYTE, self._schema_id))
+            # write the record index to the buffer
```

Comment:
you're not writing the schema id.

Reply:
(you need to, it's part of the serialization format)

Reply:
>fo.write(struct.pack('>bI', _MAGIC_BYTE, self._schema_id))

Reply:
oh, right, blind.